### ISSUE TITLE
Prepend NSI Info for Operators (of ATMs, Charging Stations, clothing bins etc)

### ISSUE THREAD

**Issue by peternewman** [streetcomplete/StreetComplete#3708]
**Use case**
> > Doing something more involved like including the NSI versions (when not duplicated), or ideally putting the NSI ones in front in a weighted way, would also be nice (given we're already using NSI for some other shop related stuff.
_Originally posted by @peternewman in https://github.com/streetcomplete/StreetComplete/issues/3690#issuecomment-1026088422_

**Proposed Solution**

> I would wait with the second until NSI has a better coverage

Looking at how the system currently behaves, I don't think that's necessary, especially if we're not switching but prepending them. If you type "S" or "Sain" in the UK, you'll find about 5 or 6 variants that are all just Sainsburys, sorted based on the current hits on the map. If you added the official (via Wikidata) via NSI first, it would just mean that the official "spelling" (should it have an apostrophe for example) would be listed first.

If you typed "Tes" for example then although Tesco isn't (currently) listed in NSI, you'd still get all the fallback Tesco entries (even if most of them are just duplicates of each other):
```
  - Tesco # 166
  - Tesco Bank # 67
  - Tesco Express # 6
  - RBS/Tesco # 2
  - Tesco Stores Ltd # 2
  - Tesco ATM # 2
  - Tesco Express... # 2
```
But at least for the entries in NSI, most of the time the consistent name would be used, so we'd slowly converge rather than continuing the divergence and requiring more clean-up in future.

I don't see any obvious downsides to doing that now (and then at some point in the future using NSI exclusively, either per country or globally.

_Originally posted by @peternewman in https://github.com/streetcomplete/StreetComplete/issues/3690#issuecomment-1026088422_

**Comment by westnordost** [comment]
Charging stations are not really in NSI (yet, except tesla), for clothing bins, there are only a few.

ATMs on the other hand, there are quite a few now. I think this was different back then. So, 👍 to that, at least for ATMs. In the midterm, we could even maybe drop the list for ATMs in favour of NSI altogether (research required).

**Comment by peternewman** [comment]
> Charging stations are not really in NSI (yet, except tesla), for clothing bins, there are only a few.

Agreed.

> ATMs on the other hand, there are quite a few now. I think this was different back then. So, 👍 to that, at least for ATMs. In the midterm, we could even maybe drop the list for ATMs in favour of NSI altogether (research required).

From looking at the list in here:
https://www.bbc.co.uk/news/business-60199157

At least some are missing currently in the UK and most of the supermarket ones aren't present either.

I've created https://github.com/osmlab/name-suggestion-index/issues/6141 to track that, but I think we may need to use the more complex solution in the interim.

> This would be the preferred solution, because it requires less code, less complexity.

If it's just suggestions, then presumably we don't have to have all of them, but there's obviously a question of how many are enough.

For now, can't we just get a list of all the suggestions from the featureDictionary (and presumably in future switch to the Place/brand style stuff where it would do smarter searching etc). But for the moment merge the two lists of suggestions?

It seems naively to me like the code to join the existing YML based array onto the bottom of the new featureDictionary one would be pretty simple?

**Comment by westnordost** [comment]
> It seems naively to me like the code to join the existing YML based array onto the bottom of the new featureDictionary one would be pretty simple?

Yeah and what to do when the user selected that suggestion? In case it is an entry from the feature dictionary, one surely wants to apply all the tags associated with the preset. While if it is from the YML, just the operator should be filled.

**Comment by peternewman** [comment]
> Yeah and what to do when the user selected that suggestion? In case it is an entry from the feature dictionary, one surely wants to apply all the tags associated with the preset. While if it is from the YML, just the operator should be filled.

I didn't think we were doing that currently (edit, we clearly are, I think I was getting mixed up between Add Place Name and Specify Shop Type; is there any reason the former doesn't handle it, I guess it may be rare to have it tagged as a bank but no name specified).

Although I certainly think it would be nice to do for the operators, getting the NSI "official" names in would be a good start, as they'll easily auto-upgrade to the official names, even if we didn't add the other tagging at the same time.

However I think there's one relatively simple solution; if you add the ability to pass in multiple brandFeatureCollections or we don't use the built in create method of osmfeatures, then we should be able to fake up basic OSM features for the YML sourced data and it will list them too.

I think the only extra code requirements (versus switching directly to NSI) are to make the fake OSM feature list from the YML and to be able to pass in a second brandFeatureCollection.

Here's an official Sainsbury's:
```
{
  "amenity/atm/sainsburysbank-396707": {
    "name": "Sainsbury's Bank",
    "locationSet": {
      "include": [
        "gb"
      ]
    },
    "geometry": [
      "point",
      "vertex"
    ],
    "matchScore": 2,
    "terms": [
      "sainsbury's bank"
    ],
    "tags": {
      "amenity": "atm",
      "brand:wikidata": "Q7400525"
    },
    "addTags": {
      "amenity": "atm",
      "brand": "Sainsbury's Bank",
      "brand:wikidata": "Q7400525",
      "brand:wikipedia": "en:Sainsbury's Bank",
      "operator": "Sainsbury's Bank",
      "operator:wikidata": "Q7400525",
      "operator:wikipedia": "en:Sainsbury's Bank"
    }
  }
}
```

Here's a fake Tesco...:
```
{
  "amenity/atm/tesco-123456": {
    "name": "Tesco",
    "locationSet": {
      "include": [
        "gb"
      ]
    },
    "geometry": [
      "point",
      "vertex"
    ],
    "matchScore": 1,
    "terms": [
      "tesco"
    ],
    "tags": {
      "amenity": "atm",
    },
    "addTags": {
      "amenity": "atm",
      "operator": "Tesco",
    }
  }
}
```

They default matchScore to 2:
https://github.com/osmlab/name-suggestion-index/blob/33e2007c7c5237f7f8afc1ac6dadc90aa8523d5f/scripts/dist_files.js#L326

So we can favour a real Sainsbury's over a faked one, we don't even need to deal with that de-duplication!

It then also provides an easy way to transition individual countries over, by just stopping generating their fake feature dictionaries.

**Comment by westnordost** [comment]
I am not going to indulge myself in an argument whether this adds complexity or not. Either you believe me or you can prove it does not by implementing it and creating a PR.

**Comment by peternewman** [comment]
I don't disagree it's more complex, but as I see it, the choice are (in rough order of complexity and also benefit):

1. Do nothing, no NSI data used, continued divergence from "official" spellings, wait until NSI has global coverage at some suitable percentage (which is likely to be a long time for some countries)
2. Switch to NSI exclusively (I'd argue this isn't complexity just a code change), may lack coverage in some countries for some aspects
3. Put the NSI names in the YAML data (as I originally proposed), probably simpler to do, we've already got code to display names, so writing them to a file shouldn't be too hard; means no additional tags from NSI are included
4. Fake more NSI data from the current YAML (as I proposed in https://github.com/streetcomplete/StreetComplete/issues/3708#issuecomment-1028563167 ). Certainly the most complex, and the complexity needs to be in SC itself, not just in some standalone thing (potentially in any language), which generates the YML file, but also has the highest benefit, should be seamless between the two, and gives options for per-country cut over.

NSI is always going to be behind live data, as it effectively relies on some tags appearing to list something, then someone linking to Wikidata (or creating if required), then a release cycle of NSI and then of SC before I can select from "My New Bank". It's really a case of how much of a gap is acceptable.

The other possible stumbling block I've just noticed, NSI collects ATM **brands** ( https://nsi.guide/index.html?t=brands&k=amenity&v=atm ), SC tags **operators**! Although clothing bins and charging are fine as-is.
